### PR TITLE
Fix ledger 191 sequence index

### PR DESCRIPTION
**Pull request by romaniy173**
Issue: Ledger sometimes failed because of 191 sequence index on OSX

### CODE REVIEW

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 1]
```diff
@@ -78,8 +79,19 @@ namespace {
       std::vector<uint8_t> buff(&buf[0], &buf[0] + 64);
```
Comment (pavel-kokolemin) [inline BlockSettleHW/ledger/ledgerDevice.cpp:79]: `buff` is not used

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 2]
```diff
@@ -1118,15 +1130,42 @@ void LedgerCommandThread::releaseDevice()
...
+      // and there is really now good solution for it, except restart dongle session
+      // till the moment error gone.
+      // https://github.com/obsidiansystems/ledger-app-tezos/blob/191-troubleshooting/README.md#error-unexpected-sequence-number-expected-0-got-191-on-macos
+      // (solution in link above is not working, given here for reference)
+      // Also it's a general issue for OSX really, but let's left it for all system, just in case
+      // And let's have 10 times threshold to avoid trying infinitively
+      static int maxAttempts = 10;
+      if (e.what() == kHidapiSequence191 && maxAttempts > 0) {
+         --maxAttempts;
+         OnExitCb([&] {
```
Comment (pavel-kokolemin) [inline BlockSettleHW/ledger/ledgerDevice.cpp:1151]: Nameless temporary variable will be destroyed right on the next line.

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 3]
```diff
@@ -1118,15 +1130,42 @@ void LedgerCommandThread::releaseDevice()
...
+         throw e;
+      }
+   }  
 }
 
-bool LedgerCommandThread::writeData(const QByteArray& input, std::string&& logHeader)
+// Do not use this function anywhere except inside exchangeData
+bool LedgerCommandThread::writeData(const QByteArray& input, const std::string& logHeader)
 {
    logger_->debug(logHeader + " - >>> " + input.toHex().toStdString());
```
[the remark this answers is not in the document]
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp:1170]: sure, fixed, thank you

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 1 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp:79]: it is in the next line

File: BlockSettleHW/hwcommonstructure.h [hunk 4]
```diff
@@ -93,4 +93,17 @@ namespace HWInfoStatus {
    const QString kCancelledByUser = QObject::tr("Cancelled by user");
 }
 
+class OnExitCb {
```
[the remark this answers is not in the document]
Comment (romaniy173) [reply BlockSettleHW/hwcommonstructure.h:96]: cool, changed

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 2 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp:1151]: true, will fix

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 1 above)
Comment (pavel-kokolemin) [reply BlockSettleHW/ledger/ledgerDevice.cpp:79]: `buf` used there, not `buff`

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 5]
```diff
@@ -67,19 +68,30 @@ namespace {
...
+      int result = hid_read(dongle, buf, Ledger::CHUNK_MAX_BLOCK);
       if (result < 0) {
          return result;
       }
 
       std::vector<uint8_t> buff(&buf[0], &buf[0] + 64);
 
       QByteArray chunk(reinterpret_cast<char*>(buf), Ledger::CHUNK_MAX_BLOCK);
-      assert(chunkNumber++ == chunk.mid(3, 2).toHex().toInt());
+      auto checkChunkIndex = [&]() {
```
Comment (sergey-chernikov) [inline BlockSettleHW/ledger/ledgerDevice.cpp:83]: Capturing everything is a bad practice - capture only used variables in a lambda.

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 6]
```diff
@@ -67,19 +68,30 @@ namespace {
...
+         auto chunkIndex = static_cast<uint16_t>(((uint8_t)chunk[3] << 8) | (uint8_t)chunk[4]);
+         if (chunkIndex != expectedChunkIndex++) {
+            if (chunkIndex == static_cast<uint16_t>(191)) {
```
Comment (sergey-chernikov) [inline BlockSettleHW/ledger/ledgerDevice.cpp:86]: Let's name this hard-coded number

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 2 above)
Comment (sergey-chernikov) [inline BlockSettleHW/ledger/ledgerDevice.cpp]: Capture list

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergey-chernikov) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: Use fmt::format syntax, not string concatenation, as you can find random symbols in format string that require variables passed to stack.

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 1 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: true! thank you

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 5 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: Fixed, but I guess this rule shouldn't be covered case for a local helpful lambda. Anyway fixed

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 6 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: Fixed, thank you

File: BlockSettleHW/ledger/ledgerDevice.cpp (hunk 2 above)
Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: deleted, thank you

File: BlockSettleHW/ledger/ledgerDevice.cpp [hunk 7]
```diff
@@ -1168,8 +1169,7 @@ bool LedgerCommandThread::writeData(const QByteArray& input, const std::string&
 {
    logger_->error(logHeader + " - >>> " + input.toHex().toStdString());
    if (sendApdu(dongle_, input) < 0) {
-      logger_->error(
-         logHeader + " - Cannot write to device.");
+      logger_->error(fmt::format("{} - Cannot write to device.", logHeader));
```
Comment (sergey-chernikov) [inline BlockSettleHW/ledger/ledgerDevice.cpp:1172]: spdlog's logger already supports fmt::format internally, so no need to put fmt::format in its methods

Comment (romaniy173) [reply BlockSettleHW/ledger/ledgerDevice.cpp]: True, fixed. Thanks

**Pull request merged**